packages/bibliography/src/main.rs

```diff
@@ -0,0 +1,358 @@
+        let mut buffer = String::new();
+        io::stdin().read_to_string(&mut buffer).unwrap();
+        serde_json::from_str(&buffer).unwrap()
+    };
+
+    match from {
+        "cite" => transform_cite(to, &input),
+        "cite-internal-do-not-use" => transform_cite_label(to, &input),
+        "bibliography" => transform_bibliography(to, &input),
+        other => {
+            eprintln!("Package does not support {other}");
+        }
+    }
+}
+
+// The name Citation clashes with Hayagriva
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct InlineCitation<'a> {
+    key: &'a str,
+    note: Option<&'a str>,
+}
+
+fn transform_cite(_to: &str, input: &Value) {
+    let citation = {
+        let key = input["data"].as_str().unwrap();
+        let note = {
+            let arg = input["arguments"]["note"].as_str().unwrap();
+            (!arg.is_empty()).then_some(arg)
+        };
+        let citation = InlineCitation { key, note };
+        serde_json::to_string(&citation).unwrap()
+    };
+
+    let output = json!([push_citation!(&citation), cite_label!(&citation)]);
+
+    println!("{output}");
+}
+
+fn transform_cite_label(_to: &str, input: &Value) {
+    let citation = input["data"].as_str().unwrap();
+
+    let labels: Vec<String> = {
+        let var = env::var("inline_citation_labels").unwrap_or("[]".to_string());
+        serde_json::from_str(&var).unwrap()
+    };
+
+    if let Some(label) = labels.iter().find_map(|s| s.strip_prefix(citation)) {
+        println!("{label}");
+    } else {
+        eprintln!("You must add a [bibliography] for [cite]s to work. If you don't want it visible, do [bibliography visibility=hidden]");
+    }
+}
+
+fn transform_bibliography(_to: &str, input: &Value) {
+    let Some(bibliography) = read_bibliography(input) else { return; };
+
+    // This is the citations (i.e [cite]s) used in the text
+    let citations: Vec<String> = {
+        let var = env::var("inline_citations").unwrap_or("[]".to_string());
+        serde_json::from_str(&var).unwrap()
+    };
+
+    // This is our database of all entries (like a db containing all biblatex-entries)
+    let mut database = Database::from_entries(bibliography.iter());
+
+    // What style to use (IEEE, APA etc)
+    let style_arg = input["arguments"]["style"].as_str().unwrap();
+
+    // Unwrap is safe since Core has checked enums already
+    let (bibliography_style, mut citation_style) = get_styles(style_arg).unwrap();
+
+    let mut output: Vec<Value> = vec![];
+
+    for citation_str in &citations {
+        // We parse the citation (which is a JSON obj)
+        let citation: InlineCitation = serde_json::from_str(citation_str).unwrap();
+
+        // We find the record by the key
+        let entry = if let Some(record) = database.records.get(citation.key) {
+            record.entry
+        } else {
+            eprintln!(
+                "Missing citation {0}, mentioned in [cite] {0} but not defined in the bibliography",
```

Comment:
Added `'` around the key and remove `[cite] ...` since we don't know how they have formatted it in the source code
```suggestion
                "Missing citation key '{0}', consider adding it to your  bibliography",
```